cpp/src/arrow/chunk_resolver.h

```diff
@@ -36,8 +39,42 @@ struct ChunkLocation {
 
   /// \brief Index of the value in the chunk
   ///
-  /// The value is undefined if chunk_index >= chunks.size()
+  /// The value is UNDEFINED if chunk_index >= chunks.size()
   int64_t index_in_chunk = 0;
+
+  /// \brief Create a ChunkLocation without asserting any preconditions.
+  static ChunkLocation Forge(int64_t chunk_index, int64_t index_in_chunk) {
+    ChunkLocation loc;
+    loc.chunk_index = chunk_index;
+    loc.index_in_chunk = index_in_chunk;
+    return loc;
+  }
+
+  ChunkLocation() = default;
+
+ public:
+  template <class ChunkResolver>
+  ChunkLocation(const ChunkResolver& resolver, int64_t chunk_index,
+                int64_t index_in_chunk)
+      : chunk_index(chunk_index), index_in_chunk(index_in_chunk) {
```

Comment:
I don't understand why we need two different constructors while this is not a public API...

Reply:
In which context do we want to produce a `ChunkLocation` out of arbitrary integers.

Reply:
I used it in tests.

```cpp
  ChunkResolver resolver(std::vector<int64_t>({0, 2, 3, 10}));  // [[0, 1], [2], [3..9]]
  ASSERT_EQ(resolver.Resolve(10).chunk_index, 3);
  ASSERT_EQ(resolver.Resolve(9), (ChunkLocation::Forge(2, 6)));
  ASSERT_EQ(resolver.Resolve(8), (ChunkLocation::Forge(2, 5)));
  ASSERT_EQ(resolver.Resolve(4), (ChunkLocation::Forge(2, 1)));
  ASSERT_EQ(resolver.Resolve(3), (ChunkLocation::Forge(2, 0)));
  ASSERT_EQ(resolver.Resolve(2), (ChunkLocation::Forge(1, 0)));
  ASSERT_EQ(resolver.Resolve(1), (ChunkLocation::Forge(0, 1)));
  ASSERT_EQ(resolver.Resolve(0), (ChunkLocation::Forge(0, 0)));
  ASSERT_EQ(resolver.Resolve(1), (ChunkLocation::Forge(0, 1)));
  ASSERT_EQ(resolver.Resolve(2), (ChunkLocation::Forge(1, 0)));
  ASSERT_EQ(resolver.Resolve(3), (ChunkLocation::Forge(2, 0)));
  ASSERT_EQ(resolver.Resolve(4), (ChunkLocation::Forge(2, 1)));
  ASSERT_EQ(resolver.Resolve(8), (ChunkLocation::Forge(2, 5)));
  ASSERT_EQ(resolver.Resolve(9), (ChunkLocation::Forge(2, 6)));
  ASSERT_EQ(resolver.Resolve(10).chunk_index, 3);
```

Reply:
Ok, can we simplify things here?
1. `ChunkLocation(int64_t, int64_t)` should simply be a regular constructor
2. the checks that are currently in the three-arg constructor are only useful if we expect arbitrary code to create their own `ChunkLocation`s. Otherwise, the checks can probably folded into `Resolve` (or perhaps even removed, if `Resolve` already guarantees that these checks are satisfied).

Reply:
I'm removing now.